ceph/nvmeof/gateway.py

```diff
@@ -0,0 +1,184 @@
+"""Ceph-NVMeoF gateway module.
+
+- Configure spdk and start spdk.
+- Configure nvme-of targets using control.cli.
+"""
+import re
+
+from ceph.ceph import CephNode
+from ceph.ceph_admin.common import config_dict_to_string
+from cli.utilities.packages import Package
+
+REPO_PATH = "/tmp/ceph-nvmeof"
+CEPH_NVMEOF_REPO = "https://github.com/ceph/ceph-nvmeof.git"
+CTRL_DAEMON = "python3 -m control"
+CTRL_DAEMON_CLI = f"{CTRL_DAEMON}.cli"
+
+
+def fetch_spdk_pid(node: CephNode):
+    """Fetch SPDK Process Id."""
+    out, _ = node.exec_command(cmd=f"pgrep -f '{CTRL_DAEMON}'", check_ec=False)
+    return out.strip() if out else False
+
+
+def kill_spdk_process(node: CephNode):
+    """Kill SPDK daemon."""
+    pid = fetch_spdk_pid(node)
+    if pid:
+        node.exec_command(cmd=f"kill -9 {pid}", check_ec=False, sudo=True)
+
+
+def configure_spdk(node: CephNode):
+    """SPDK installation on Gateway node.
+
+    - Supports only RHEL-9.
+
+    Args:
+        node: Ceph Node object.
+    """
+    pkg_mgr = Package(nodes=[node])
+    pkg_mgr.install("git")
+
+    # Clone the ceph-nvmeof repository
+    node.exec_command(
+        sudo=True,
+        cmd=f"rm -rf {REPO_PATH}; git clone --recursive {CEPH_NVMEOF_REPO} {REPO_PATH}",
+    )
+    out, _ = node.exec_command(cmd=f"cat {REPO_PATH}/ceph-nvmeof.conf")
+
+    def update_cfg(cfg, key, value):
+        return re.sub(f"{key} = (.*)", f"{key} = {value}", cfg)
+
+    conf = update_cfg(out, "spdk_path", REPO_PATH)
+    conf = update_cfg(conf, "addr", node.ip_address)
+
+    _file = node.remote_file(
+        sudo=True, file_name=f"{REPO_PATH}/ceph-nvmeof.conf", file_mode="w"
+    )
+    _file.write(conf)
+    _file.flush()
+
+    # Install pre-requisites
+    pkg_mgr.install(
+        "https://dl.fedoraproject.org/pub/epel/epel-release-latest-9.noarch.rpm"
+    )
+    node.exec_command(
+        sudo=True,
+        cmd="subscription-manager repos --enable=codeready-builder-for-rhel-9-x86_64-rpms",
+    )
+    pkg_mgr.install(
+        "CUnit-devel libiscsi-devel json-c-devel libcmocka-devel python-rados librbd-devel"
+    )
+    node.exec_command(
+        sudo=True,
+        cmd="sh -c 'echo 4096 > /sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages'",
+    )
+    if node.exec_command(
+        sudo=True,
+        cmd=f"cd {REPO_PATH}/spdk; ./scripts/pkgdep.sh; ./configure --with-rbd; make",
+        long_running=True,
+        check_ec=False,
+    ):
+        raise Exception("SPDK pre-requisites installation failed...")
+    node.exec_command(
+        sudo=True,
+        cmd=f"cd {REPO_PATH}; make grpc; nohup {CTRL_DAEMON} > output.log 2>&1 &",
+    )
+
+    if not fetch_spdk_pid(node):
+        raise Exception("Control and SPDK process is not running.")
+
+
+def run_control_cli(node: CephNode, action, **cmd_args):
+    """Run CLI via control daemon."""
+    return node.exec_command(
+        sudo=True,
+        cmd=f"cd {REPO_PATH}; {CTRL_DAEMON_CLI} {action} {config_dict_to_string(cmd_args)}",
```

Comment:
IMO, It would be good to log output of each command, so we have them in logs

Reply:
Its good to have this message in the log - INFO:control.server:Starting server instead of full ouput

Reply:
Updated.